src/software/ai/navigator/obstacle/obstacle_factory_test.cpp

```diff
@@ -31,16 +32,21 @@ TEST_F(ObstacleFactoryTest, create_rectangle_obstacle)
     Rectangle rectangle(Point(1, 3), Point(5, 8));
     Polygon expected(Rectangle(Point(.883, 2.883), Point(5.117, 8.117)));
     ObstaclePtr obstacle = obstacle_factory.createObstacleFromRectangle(rectangle);
-
-    try
-    {
-        auto polygon_obstacle = dynamic_cast<PolygonObstacle&>(*obstacle);
-        EXPECT_EQ(expected, polygon_obstacle.getPolygon());
-    }
-    catch (std::bad_cast)
-    {
-        ADD_FAILURE() << "PolygonObstaclePtr was not created for a rectangle";
-    }
+    ObstacleShape obstacle_shape = obstacle->getObstacleShape();
+
+    std::visit(overload{[&obstacle_shape, expected](const ConvexPolygon& convex_polygon) {
+                            EXPECT_TRUE(::Test::TestUtil::equalWithinTolerance(
+                                expected, convex_polygon, METERS_PER_MILLIMETER));
+                        },
+                        [&obstacle_shape, expected](const Circle& circle) {
+                            ADD_FAILURE()
+                                << "Expected ConvexPolygon, Obstacle was of type Circle";
+                        },
+                        [](auto arg) {
```

Comment:
Ah I didn't realize when you were talking about dynamic_cast vs visitor it was in the context of tests. For tests I actually prefer `dynamic_cast` because we only care about checking for one of the possible outputs, and everything else is a failure. We don't need to handle every type, so IMO visitor is overkill here.

@garethellis0 thoughts?

Reply:
I still like the visitor design pattern here, we should just explicitly fail on the other cases IMO.

Reply:
devil's advocate, we use `dynamic_cast` with try-catch everywhere else in our tests (like action tests, tactic tests etc.). Can you be more specific as to why you prefer this?

Reply:
You know what, you're right. Looking at other places where we've used the `dynamic_cast` - `try-catch` pattern, I think that would be totally fine here.

Reply:
Since I made a visitor for drawing/navigator, I might as well use it here